File: tests/undef_scan.py
"""index.html の <script> から、宣言が見つからない大文字識別子を洗い出す。
STAGE_H のような「実行されるまで気付けない未定義参照」を静的に拾うのが目的。"""
import io, re, sys

path = 'C:/Users/futsa/OneDrive/デスクトップ/業務効率化/カタモン/index.html'
src = io.open(path, encoding='utf-8').read()
script = re.findall(r'<script(?![^>]*src=)[^>]*>(.*?)</script>', src, re.S)[0]

# コメントと文字列リテラルを潰してから走査する
body = re.sub(r'//[^\n]*', '', script)
body = re.sub(r'/\*.*?\*/', '', body, flags=re.S)
for pat in [r'`(?:[^`\\]|\\.)*`', r"'(?:[^'\\\n]|\\.)*'", r'"(?:[^"\\\n]|\\.)*"']:
    body = re.sub(pat, ' "" ', body)

ident = r'\b[A-Z][A-Z0-9_]{2,}\b'
used = set(re.findall(ident, body))
declared = set(re.findall(r'(?:const|let|var|function|class)\s+([A-Z][A-Z0-9_]{2,})\b', body))
# 分割代入とプロパティ参照は別物なので除外する
declared |= set(re.findall(r'([A-Z][A-Z0-9_]{2,})\s*(?:=|,)', body)) & used
props = set(re.findall(r'\.\s*([A-Z][A-Z0-9_]{2,})\b', body))
keys = set(re.findall(r'([A-Z][A-Z0-9_]{2,})\s*:', body))
builtin = {'NaN', 'JSON', 'URL', 'GET', 'PUT', 'POST', 'DELETE'}

missing = sorted(used - declared - props - keys - builtin)
print('宣言が見つからない大文字識別子:')
if not missing:
    print('  なし')
for m in missing:
    lines = [i + 1 for i, line in enumerate(script.split('\n')) if re.search(r'\b' + m + r'\b', line)]
    print('  %-26s 出現 %d 箇所  script内行: %s' % (m, len(lines), lines[:6]))
sys.exit(1 if missing else 0)

File: tests/undef-scan.js
const assert = require('node:assert/strict');
const fs = require('node:fs');
const path = require('node:path');

const IDENTIFIER = /\b[A-Z][A-Z0-9_]{2,}\b/g;
const BUILT_INS = new Set(['NaN', 'JSON', 'URL', 'GET', 'PUT', 'POST', 'DELETE']);

function findMissingUppercaseIdentifiers(script) {
  let body = script
    .replace(/\/\/[^\n]*/g, '')
    .replace(/\/\*[\s\S]*?\*\//g, '');
  for (const pattern of [
    /`(?:[^`\\]|\\.)*`/g,
    /'(?:[^'\\\n]|\\.)*'/g,
    /"(?:[^"\\\n]|\\.)*"/g,
  ]) body = body.replace(pattern, ' "" ');

  const used = new Set(body.match(IDENTIFIER) || []);
  const declared = new Set([...body.matchAll(/(?:const|let|var|function|class)\s+([A-Z][A-Z0-9_]{2,})\b/g)].map((match) => match[1]));
  for (const match of body.matchAll(/([A-Z][A-Z0-9_]{2,})\s*(?:=|,)/g)) {
    if (used.has(match[1])) declared.add(match[1]);
  }
  const properties = new Set([...body.matchAll(/\.\s*([A-Z][A-Z0-9_]{2,})\b/g)].map((match) => match[1]));
  const objectKeys = new Set([...body.matchAll(/([A-Z][A-Z0-9_]{2,})\s*:/g)].map((match) => match[1]));
  return [...used].filter((name) => !declared.has(name) && !properties.has(name) && !objectKeys.has(name) && !BUILT_INS.has(name)).sort();
}

const html = fs.readFileSync(path.join(__dirname, '..', 'index.html'), 'utf8');
const inlineScript = /<script(?![^>]*\bsrc=)[^>]*>([\s\S]*?)<\/script>/i.exec(html)?.[1];
assert.ok(inlineScript, 'index.html にインラインのゲームscriptが必要です。');

const missing = findMissingUppercaseIdentifiers(inlineScript);
assert.deepEqual(missing, [], `宣言が見つからない大文字識別子: ${missing.join(', ')}`);
assert.deepEqual(
  findMissingUppercaseIdentifiers('const DECLARED = 1; MISSING_TOKEN;').filter((name) => name === 'MISSING_TOKEN'),
  ['MISSING_TOKEN'],
  '検査器自身が未定義識別子を検出できること',
);

console.log('未定義大文字識別子: なし（2/2 passed）');
